# Linear momentum balance

## I. Governing theory

For a fixed control volume in an inertial reference frame, the integral linear-momentum equation is,

$$ \sum \mathbf{F}_\text{ext} = \frac{\partial}{\partial t}\int_{\text{CV}} \rho \mathbf{V} dV + \int_{\text{CS}} \rho \mathbf{V}(\mathbf{V} \cdot \mathbf{n}) dA$$

### Induced process
Let the momentum $\mathbf{P}_\text{sys}$ , 
$$\mathbf{P}_\text{sys} = \int_{\text{sys}} \mathbf{V} dm $$

Because of Newton's second law,
$$ \sum \mathbf{F}_\text{ext} = \frac{d \mathbf{P}_\text{sys}}{dt}$$

Applying the Reynolds transport theorem gives,

$$\frac{D B_{\text{sys}}}{Dt} = \frac{\partial}{\partial t} \int_{\text{CV}} \rho b \, dV + \int_{\text{CS}} \rho b (\mathbf{V}_{\text{r}} \cdot \mathbf{n}) \, dA$$


It we chose $B_{\mathrm{sys}}$ is momentum, We can write B to $$B_{\mathrm{sys}} = \mathbf P_{\mathrm{sys}} = \int_{\mathrm{sys}}\mathbf V \,dm$$
Therefore, $$b=\frac{dB}{dm}=\mathbf V$$

So, 

$$ \sum \mathbf{F}_\text{ext} = \frac{\partial}{\partial t}\int_{\text{CV}} \rho \mathbf{V} dV + \int_{\text{CS}} \rho \mathbf{V}(\mathbf{V} \cdot \mathbf{n}) dA = \frac{D \mathbf{P}_\text{sys}}{Dt} $$

To organize this, we can get 

$$ \sum \mathbf{F}_\text{ext} = \frac{\partial}{\partial t}\int_{\text{CV}} \rho \mathbf{V} dV + \int_{\text{CS}} \rho \mathbf{V}(\mathbf{V} \cdot \mathbf{n}) dA$$

### Interpretation 

- $ \sum \mathbf{F}_\text{ext} $ : External forces that affect the fluid in the control volume.

The net external force may include the following contributions.
|Force|Expression or meaning|
|:---:|:---:|
|pressure force| $-\int_{\text{CS}} p \, \mathbf{n} dA$|
|Wall or surface force| Resultant force exerted by the solid boundary on the fluid|
|Gravity force| $\int_{\text{CV}} \rho \, \mathbf{g} dV$|
|shear force| Tangential component of the surface force |
|Support or reaction force| Unknown external force determined from the momentum equation|

The shear force is the tangential component of the total wall or surface force.

- $ \frac{\partial}{\partial t}\int_{\text{CV}} \rho \mathbf{V} dV $ : Rate of change of linear momentum in the control volume.

- $ \int_{\text{CS}} \rho \mathbf{V}(\mathbf{V} \cdot \mathbf{n}) dA $ : Net rate of momentum transported through the control surface.
------------------------
$ \mathbf{V} \cdot \mathbf{n} < 0 $ : at an inlet.

$ \mathbf{V} \cdot \mathbf{n} > 0 $ : at an outlet.

So, the fluid flow is one-dimensional and steady, and we get
$$ \sum \mathbf{F}_\text{ext} = \sum_\text{out} \dot{m}\mathbf{V}_\text{out} - \sum_\text{in} \dot{m}\mathbf{V}_\text{in} $$

### Non-uniform Velocity Distribution.
For a nonuniform velocity distribution, a momentum-flux correction factor $\beta$ is introduced.

$$\beta = \frac{1}{AV^2_\text{avg}} \int_{A} u^2 dA$$

Therefore, for steady one-dimensional flow,
$$ \sum \mathbf{F}_\text{ext} = \sum_\text{out} \beta \, \dot{m}\mathbf{V}_\text{avg} -  \sum_\text{in} \beta \, \dot{m}\mathbf{V}_\text{avg} $$

+) Each inlet or outlet may have its own value of

$\beta$ : momentum-flux correction factor. 
When fluid has uniform velocity distribution, $\beta=1$

### moving control volume

When the control volume is moving, fluid has relative velocity, $\mathbf{V}_r$
$$\mathbf{V}_r = \mathbf{V} - \mathbf{V}_\text{CV}$$

so, we have to distinguish 
$$\rho\mathbf{V} \;\; \text{vs} \;\; \mathbf{V}_r \cdot \mathbf{n} $$

Because, 
$\rho\mathbf{V}$ is a momentum on transformation, and 

$ \mathbf{V}_r \cdot \mathbf{n} $ is a fluid's velocity that passes through the cross area.

Absolute velocity determines the momentum carried by the fluid, whereas relative velocity determines whether and how rapidly the fluid crosses the moving control surface.

### Noninertial reference Frame
When the coordinate system has translation acceleration $\mathbf{a}_0$, 	virtualacceleration per kilogram is $$-\mathbf{a}_0$$
Therefore, Inertial is 
$$\mathbf{F}_\text{inertial} = -\int_{\text{CV}} \rho \mathbf{a}_0 dV$$

In an inertial frame of reference, let's write the location vector $\mathbf{S}_i$
$$\mathbf{S}_i = \mathbf{r} + \mathbf{R}$$

absolute velocity $\mathbf{V}_{\mathrm{abs}}$ is $\frac{d\mathbf{S}_i}{dt}$,
$$\mathbf{V}_{\mathrm{abs}} = \mathbf{V}_{O} + \frac{d\mathbf{R}}{dt} + \mathbf{\Omega}\times\mathbf{r}$$

or 

$$ \mathbf{V}_{\mathrm{abs}} = \mathbf{V}_{O} + \mathbf{V}_{\mathrm{rel}} + \mathbf{\Omega}\times\mathbf{r}. $$

absolute acceleration $\mathbf{a}_{\mathrm{abs}}$ is $\frac{d\mathbf{V}_i}{dt}$,
$$\mathbf{a}_{\mathrm{abs}} = \frac{D\mathbf{V}}{Dt} + \frac{d^2\mathbf{R}}{dt^2} + \frac{d\mathbf{\Omega}}{dt}\times\mathbf{r} + (2\mathbf{\Omega}\times\mathbf{V}) + \mathbf{\Omega}\times(\mathbf{\Omega}\times\mathbf{r}) $$

or 

$$\mathbf{a}_{\mathrm{abs}} = \mathbf{a}_{O} + \mathbf{a}_{\mathrm{rel}} +
2\mathbf{\Omega}\times\mathbf{V}_{\mathrm{rel}} + \dot{\mathbf{\Omega}}\times\mathbf{r} +
\mathbf{\Omega}\times\left(\mathbf{\Omega}\times\mathbf{r}\right).
$$


$\mathbf{R}$ is the absolute position of the moving reference-frame origin, and r is the position relative to that origin, $\mathbf{\Omega}$ is angular velocity

### Interpretation 

- $\frac{d^2\mathbf{R}}{dt^2}$ : starting point of noninertial reference Frame's acceleration
- $\frac{d\mathbf{\Omega}}{dt}\times\mathbf{r}$ : angular acceleration
- $2\mathbf{\Omega}\times\mathbf{V}$ : Coriolis acceleration
- $\mathbf{\Omega}\times(\mathbf{\Omega}\times\mathbf{r})$ : centripetal acceleration term in the absolute-acceleration transformation. When moved to the force side of an equation written in the rotating frame, its negative corresponds to the centrifugal inertial-force term.

||moving control volume|noninertial reference frame|
|:---:|:---:|:---:|
|core|control surface is moving|Coordinate system has acceleration|
|flux|relative velocity|relative velocity|
|momentum definition|absolute velocity|have to translate|
|Inertial force| depends on case |need virtual force|
|Example|CV follows the nozzle that moves|fluid in the car that has acceleration|

## II. Computational applications

## Case 1: Straight duct with unchanged velocity

### Goal

Demonstrate the function can calculate the $$\sum_{\text{out}}\dot{m}\mathbf{V} - \sum_{\text{in}}\dot{m}\mathbf{V}$$ correctly.

### Assumptions and given conditions

1. The flow is steady.
2. The fluid is incompressible.
3. There is one inlet and one outlet.
4. The inlet and outlet areas are equal.
5. Steady incompressible continuity therefore gives
   $\mathbf{V}_{\mathrm{in}}=\mathbf{V}_{\mathrm{out}}$.
4. $\,\dot{m}=4.0\ \mathrm{kg/s}$, $\mathbf{V}_{\text{in}}=(3.0,0,0)\ \mathrm{m/s}$, and $\mathbf{V}_{\text{out}}=(3.0,0,0)\ \mathrm{m/s}$.



### We have to get,

1. net momentum flux vector, $\mathbf{T}_\text{net}$
2. Why does the $\mathbf{T}_\text{net}$ become zero?
3. Why can pressure force and frictional force exist when $\mathbf{T}_\text{net} = 0$

In [1]:
from src.momentum_tool import(
  mass_flow_rate,
    net_momentum_flux,
)
import numpy as np

In [2]:
# case 1
mdot = 4.0

inlets = [{"velocity": (3.0, 0, 0), "mass_flow_rate": mdot}]
outlets = [{"velocity": (3.0, 0, 0), "mass_flow_rate": mdot}] # [m/s]

p_net = net_momentum_flux(inlets, outlets)

print(f" P_net = {p_net}")

 P_net = [0. 0. 0.]


### Results

$$\mathbf{T}_\text{net} = [0. 0. 0.]\, \mathrm{N}$$

### Physical Interpretation

비압축성 유체가 일정한 단면을 가진 검사체적에서 직선 정상유동을 할 때, 검사체적에 출입하는 유체의 단면 평균 속도는 같고,  $$\int_{CS} \rho \mathbf{V}(\mathbf{V} \cdot \mathbf{n}) dA$$
으로 나타내는 순 운동량 플럭스가 $0$이 된다.

When an incompressible fluid undergoes steady, straight flow through a control volume with a constant cross-sectional area, the cross-sectional average velocities of the fluid entering and leaving the control volume are the same. Therefore, the net momentum flux expressed as

$$
\int_{\mathrm{CS}}
\rho\mathbf{V}(\mathbf{V}\cdot\mathbf{n})\,dA
$$

becomes zero.

### Answer the questions
Q1. Why does the $\mathbf{T}_\text{net}$ become zero?

A1. 정상 유동 아래 검사체적의 질량은 보존되기에 유입되는 질량과 유출되는 질량이 같고, 일정한 단면적을 가지기에 입구와 출구의 유속 또한 동일하다. 따라서 $\sum_{\text{out}}\dot{m}\mathbf{V} - \sum_{\text{in}}\dot{m}\mathbf{V}$으로 계산되는 순 운동량 플럭스는 0이 된다.

Under steady flow, mass is conserved within the control volume, so the mass entering and leaving the control volume is the same. Since the inlet and outlet also have the same cross-sectional area, their velocities are the same. Therefore, the net momentum flux calculated by

$$
\sum_{\text{out}}\dot{m}\mathbf{V}-\sum_{\text{in}}\dot{m}\mathbf{V}
$$

becomes zero.

------------------------------------------------------

Q2. Why can pressure force and frictional force exist when $\mathbf{T}_\text{net} = 0$

A2. 레이놀즈 수송정리와 뉴턴의 제 2법칙을 활용하여 선형 운동량 방정식을 도출할 때,

   $$\sum \vec{F}_\text{c.v} = \frac{\partial}{\partial t} \int_{\text{CV}} \vec{V} \rho \, dV + \int_{\text{CS}} \vec{V} \rho(\mathbf{V} \cdot \mathbf{n}) \, dA $$

을 얻게되고, 정상 유동의 경우 

$$\sum \vec{F}_\text{c.v} = \int_{\text{CS}} \vec{V} \rho(\mathbf{V} \cdot \mathbf{n}) \, dA $$
을 얻는다. 이때 우변이 0이 되더라도 좌변은 알짜힘이기에, 압력력과 마찰력의 벡터합이 0이 된다면 충분히 압력력과 마찰력이 존재할 수 있다.

When the linear momentum equation is derived using the Reynolds transport theorem and Newton's second law, we obtain

$$\sum\mathbf{F}_{\mathrm{CV}}=\frac{\partial}{\partial t}
\int_{\mathrm{CV}}\rho\mathbf{V}\,dV
+
\int_{\mathrm{CS}}
\rho\mathbf{V}(\mathbf{V}\cdot\mathbf{n})\,dA.
$$

For steady flow, this becomes

$$
\sum\mathbf{F}_{\mathrm{CV}}=
\int_{\mathrm{CS}}
\rho\mathbf{V}(\mathbf{V}\cdot\mathbf{n})\,dA.
$$

Even if the right-hand side is zero, the left-hand side represents the net force. Therefore, pressure and friction forces can still exist as long as their vector sum is zero.


## Case 2: Accelerating nozzle flow

### Goal

Check the relation between speed and momentum flux.

### Assumptions and given conditions

1. The flow is steady.
2. The fluid is incompressible.
3. The nozzle is horizontal
4. The fluid flows in the positive x-direction.
5. $\rho = 1000 \mathrm{kg/m^3} $
6. inlet : $A_1 = 0.010 \mathrm{m^2}, \mathbf{V}_1 = 2.0 \mathrm{m/s}$
7. outlet: : $A_2 = 0.004 \mathrm{m^2} $
   

### We have to get,

1. mass flow rate, $\dot{m}$
2. outlet velocity $\mathbf{V}_2$
3. inlet momentum flux vector, $\mathbf{T}_\text{in}$
4. outlet momentum flux vector, $\mathbf{T}_\text{out}$
5. net momentum flux vector, $\mathbf{T}_\text{net}$

### Question

1. Why is the outlet speed bigger than the inlet speed?
2. Is the x-component of the net momentum flux positive?
3. What direction of external force does this result indicate?
4. Why do we have to consider pressure force to obtain the actual force?

In [3]:
# case 2

rho = 1000 # [kg/m^3]

A_1 = 0.010 # [m^2]
V_1 = 2.0 # [m/s]

A_2 = 0.004 # [m^2]

V_2 = A_1*V_1/A_2

mdot = mass_flow_rate(rho, A_1, V_1)

inlets = [ {"velocity": (V_1, 0, 0), "mass_flow_rate": mdot} ]
outlets = [ {"velocity": (V_2, 0, 0), "mass_flow_rate": mdot} ]

net = net_momentum_flux(inlets, outlets)

print(f"mass flow rate: {mdot}, V_2: {V_2}, inlet_momentum_flux: {V_1*mdot}, outlet_momentum_flux: {V_2*mdot}, net_momentum_flux: {net}")

mass flow rate: 20.0, V_2: 5.0, inlet_momentum_flux: 40.0, outlet_momentum_flux: 100.0, net_momentum_flux: [60.  0.  0.]


### Results

$$
\dot{m}=20.0\ \mathrm{kg/s},
\qquad
\mathbf{V}_2=(5.0,0,0)\ \mathrm{m/s}
$$

$$
\mathbf{T}_{\text{in}}=(40.0,0,0)\ \mathrm{N},
\qquad
\mathbf{T}_{\text{out}}=(100.0,0,0)\ \mathrm{N}
$$

$$
\mathbf{T}_{\text{net}}=(60.0,0,0)\ \mathrm{N}
$$

### Physical Interpretation

속도의 크기가 변하면 $$\int_{\text{CS}} \vec{V} \rho(\mathbf{V} \cdot \mathbf{n}) \, dA $$로 정의되는 운동량 플럭스에 변화가 생긴다. 정상 유동 아래 질량 유량은 유입량과 유출량이 동일하기에, 운동량 플럭스의 크기는 출입구의 속도에 비례한다. 출구의 단면적이 입구의 단면적 보다 작다면, 출구의 유속은 입구의 유속보다 크고, 이는 출구 운동량 플럭스가 입구 운동량 플럭스 보다 큼을 의미하며, x축 양의 방향으로 움직이는 case 2의 경우, 순운동량 플럭스가 양의 값을 가진다.

When the magnitude of velocity changes, the momentum flux defined by

$$
\int_{\mathrm{CS}}
\rho\mathbf{V}(\mathbf{V}\cdot\mathbf{n})\,dA
$$

also changes. Under steady flow, the inlet and outlet mass flow rates are the same, so the magnitude of the momentum flux is proportional to the velocity at each opening. If the outlet area is smaller than the inlet area, the outlet velocity is greater than the inlet velocity. This means that the outlet momentum flux is greater than the inlet momentum flux. Since the flow in Case 2 moves in the positive $x$-direction, the net momentum flux has a positive $x$-component.


### Answer the questions
Q1. Why is the outlet speed bigger than the inlet speed?

A1. 비압축성 정상 유동하에서 검사체적의 질량은 보존되므로, 질량 유입률과 질량 유출률은 동일해야 한다. 따라서 $A_1V_1=A_2V_2$을 만족시켜야 하며 이로부터 $V_2=A_1V_1/A_2$를 얻는다. 

$\frac{A_1}{A_2}>1$이기에, $V_2>V_1$의 결과를 얻는다.

Under steady incompressible flow, mass is conserved within the control volume, so the inlet and outlet mass flow rates must be the same. Therefore,

$$
A_1V_1=A_2V_2
$$

must be satisfied, and this gives

$$
V_2=\frac{A_1V_1}{A_2}.
$$

Since $\frac{A_1}{A_2}>1$, we obtain $V_2>V_1$.

--------------------------------------

Q2. Is the x-component of the net momentum flux positive?

A2. 그렇다. $\mathbf{T}_\text{net}: [60.  0.  0.]$에서 확인할 수 있다.

Yes. This can be confirmed from

$$
\mathbf{T}_{\text{net}}=(60.0,0,0)\ \mathrm{N}.
$$

--------------------------------------

Q3. What is the external force's direction that this result means?

A3. $$\mathbf{T}_\text{net}: [60.  0.  0.]$$이기에, 

$$\sum \vec{F}_\text{c.v} = \int_{\text{CS}} \vec{V} \rho(\mathbf{V} \cdot \mathbf{n}) \, dA $$
에 따라 $x+$ 방향의 외력이 필요함을 의미한다.

Since

$$
\mathbf{T}_{\text{net}}=(60.0,0,0)\ \mathrm{N},
$$

the steady-flow momentum equation

$$
\sum\mathbf{F}_{\mathrm{CV}}=
\int_{\mathrm{CS}}
\rho\mathbf{V}(\mathbf{V}\cdot\mathbf{n})\,dA
$$

indicates that a net external force in the positive $x$-direction is required.

--------------------------------------

Q4. Why do we have to consider the pressure force to get the real force?

A4. 선형 운동량 방정식 $$\sum \vec{F}_\text{c.v} = \int_{\text{CS}} \vec{V} \rho(\mathbf{V} \cdot \mathbf{n}) \, dA $$의 $\sum \vec{F}_\text{c.v}$은 검사체적에 작용하는 모든 외력의 합으로, 검사체적의 출입구에 수직으로 작용하는 압력력 등의 기타 모든 외력들을 고려해야 한다.

In the linear momentum equation, $\sum\mathbf{F}_{\mathrm{CV}}$ is the sum of all external forces acting on the control volume. Therefore, the pressure forces acting normal to the inlet and outlet must also be considered when determining the actual wall or support force.

## Case 3: $90^\circ$ elbow with pressure forces

### Goal

Check the effect of a change in flow direction on momentum conservation.

### Assumptions and given conditions

1. The flow is steady.
2. The fluid is incompressible.
3. The pressure is gauge pressure.
4. $\, \rho = 1000\mathrm{kg/m^3}$
5. inlet : $ \text{x-axis}, A_1 = 0.005 \mathrm{m^2}, \mathbf{V}_1 = 4.0 \mathrm{m/s}, p_1 = 120 \mathrm{kPa}$
6. outlet: : $\text{y-axis}, A_2 = 0.005 \mathrm{m^2}, \mathbf{V}_2 = 4.0 \mathrm{m/s}, p_2 = 80 \mathrm{kPa} $
7. Neglect the fluid weight.
8. Do not model viscous shear separately.
9. Represent the total mechanical interaction between the elbow wall
   and the fluid by the resultant wall force
   $\mathbf F_{\mathrm{wall}}$.
### We have to get,

1. mass flow rate, $\dot{m}$
2. net momentum flux vector, $\mathbf{T}_\text{net}$
3. inlet pressure force vector, $\mathbf{F}_\text{pressure,in}$
4. outlet pressure force vector, $\mathbf{F}_\text{pressure,out}$
5. wall force on the fluid, $\mathbf{F}_\text{wall}$
6. Force exerted by the fluid on the elbow, $\mathbf{F}_\text{fluid}$

In [4]:
# case 3

rho = 1000 # [kg/m^3]

P_1 = 120000 # [Pa]
A_1 = 0.005 # [m^2]
V_1 = 4.0 # [m/s]

P_2 = 80000 # [Pa]
A_2 = 0.005 # [m^2]
V_2 = 4.0 # [m/s]

F_1 = P_1*A_1
F_2 = P_2*A_2

m_flow_rate = rho*A_1*V_1

inlets = [ {"velocity": (V_1, 0, 0), "mass_flow_rate": m_flow_rate}]
outlets = [ {"velocity": (0, V_2, 0), "mass_flow_rate": m_flow_rate}]

net = net_momentum_flux(inlets, outlets)

F_pressure_in = np.array([F_1, 0, 0], dtype=float)
F_pressure_out = np.array([0, -F_2, 0], dtype=float)

F_wall = net - F_pressure_in - F_pressure_out
F_fluid = -F_wall

print(f"mass flow rate: {m_flow_rate}, T_net: {net}, F_in: {F_pressure_in}, F_out: {F_pressure_out}, F_wall: {F_wall}, F_fluid: {F_fluid}")

mass flow rate: 20.0, T_net: [-80.  80.   0.], F_in: [600.   0.   0.], F_out: [   0. -400.    0.], F_wall: [-680.  480.    0.], F_fluid: [ 680. -480.   -0.]


### Results

$$ \dot{m}: 20.0 \, \mathrm{kg/s}$$ $$\mathbf{T}_\text{net}: [-80.  80.   0.] \, \mathrm{N}$$ 
$$\mathbf{F}_\text{in}: [600.   0.   0.] \, \mathrm{N}, \, \mathbf{F}_\text{out}: [0. -400. 0.] \, \mathrm{N}, $$
$$\mathbf{F}_\text{wall}: [-680. 480. 0.] \, \mathrm{N}, \, \mathbf{F}_\text{fluid}: [680. -480. 0.] \, \mathrm{N}$$

### Physical Interpretation
입구와 출구의 속력은 같으나, 방향이 변하기에 순 운동량 플럭스는 0이 아니다. 각각 $+x, -y$ 방향으로 입출구에 수직으로 작용하는 압력과 외력을 성분 분해하여 외력의 방향별 성분을 쉽게 구할 수 있다. 압력력만으로는 case 3에 부합하는 운동량 변화를 만들 수 없기에, 벽이 추가적인 외력을 유체에 가해주어야 한다. 마찰력과 중력을 무시하는 case 3의 경우, 벽이 유체에 가하는 힘을 알면 작용-반작용의 원리에 의해 유체가 벽에 가하는 힘을 쉽게 알 수 있다.

The inlet and outlet speeds are the same, but the net momentum flux is not zero because the direction of the velocity changes. The pressure forces acting normal to the inlet and outlet point in the positive $x$- and negative $y$-directions, respectively. By resolving the pressure and external forces into directional components, each component of the external force can be obtained more easily.

The pressure forces alone cannot produce the momentum change required in Case 3, so the wall must exert an additional external force on the fluid. Since friction and gravity are neglected in this case, once the force exerted by the wall on the fluid is known, the force exerted by the fluid on the wall can be found directly from the action-reaction principle.


## III. Conclusion and limitations

### Conclusion
단면적이 일정한 직선 정상 유동인 경우 동일한 입출구 속도를 가지며 순 운동량 플럭스가 0이다.
노즐의 경우 속력이 증가하면 양의 x축 방향 운동량의 변화가 발생한다. 
방향이 변화하는 경우 같은 입출구 면적 및 속도를 가지더라도 운동량의 변화가 발생한다.
순 운동량 플럭스는 합력을 의미하며, 일반적으로 개별 외력에 직접 대응되지는 않는다.
적분적 해석은 전체의 힘을 구할 수 있으나, 내부적 분포에 대한 정보는 제공하지 않는다.

For steady, straight flow through a constant-area duct, the inlet and outlet velocities are the same, so the net momentum flux is zero. In the nozzle, an increase in speed produces a change in momentum in the positive $x$-direction. When the flow direction changes, momentum also changes, even if the inlet and outlet areas and speeds are the same. The three cases demonstrate that linear momentum must be treated as a vector quantity. A flow may have zero net momentum flux when the inlet and outlet momentum vectors are equal, while changes in either speed or direction produce a nonzero momentum-flux difference. The momentum equation determines the resultant external force, but individual pressure, wall, gravity, and support forces must be identified separately.

### Limitations

지금의 모델은 입출구 평균 속도에 따른 운동량 변화, 검사체적에 작용하는 합력, 특정 지점에 작용하는 반력, 압력에 의한 힘과 벽면에 의한 힘, 움직이는 검사체적의 상대속도에 대해 분석할 수 있다. 이는
정상유동, 균일한 속도분포, 비압축성 유체 및 중력과 마찰을 무시하는 모델을 설정하였기 때문이다. 따라서 추가적인 항을 요구하는 포괄적 유체 분석을 수행하기에는 아직 어려움이 따른다. 현재는 관 내부의 전체 속도장, 벽면의 속도 구배, 위치별 전단 응력, 국소적 압력 구배, 점성에 의한 속도 분포 등을 분석하기 어렵다.

The current model can calculate momentum-flux changes based on
cross-sectional average velocities, resultant forces on a control
volume, pressure-force contributions, wall reaction forces, and the
relative velocity associated with a moving control surface.

However, the computational implementation is primarily restricted to
steady, incompressible, one-dimensional flows with uniform velocity
distributions. It does not yet determine the complete internal velocity
field, local velocity gradients, local shear-stress distributions,
spatial pressure gradients, or the formation of viscous velocity
profiles.

A complete moving-control-volume or non-inertial momentum solver has
also not yet been implemented. The next project stage will therefore use
differential conservation equations to describe momentum variation
within the flow domain.

-----------------------------------------

Integral linear-momentum analysis determines the resultant force and the overall momentum transfer across a control volume. However, it does not provide the local velocity and shear-stress distributions inside the flow. The next stage therefore introduces differential conservation equations to describe how momentum varies from point to point.

## IV. Verification

The implementation is verified using four types of tests:

1. benchmark tests with manually calculated results;
2. limiting-case tests;
3. invalid-input tests;
4. conservation-residual tests.

Passing a test does not validate the physical assumptions of the model.
It verifies that the code reproduces the intended mathematical model.

In [5]:
import numpy as np
from src.momentum_tool import (mass_flow_rate, net_momentum_flux, relative_velocity)

In [6]:
def run_test(test_function):
    try:
        test_function()
        print(f"PASS: {test_function.__name__}")
        return True

    except AssertionError as error:
        print(f"FAIL: {test_function.__name__}")
        print(f"      {error}")
        return False

    except Exception as error:
        print(f"ERROR: {test_function.__name__}")
        print(f"       Unexpected {type(error).__name__}: {error}")
        return False

In [7]:
def test_mass_flow_rate_normal_case():
    rho = 1000.0       # kg/m^3
    area = 0.02        # m^2
    normal_speed = 3.0 # m/s

    calculated = mass_flow_rate(rho, area, normal_speed)
    expected = 60.0

    assert np.isclose(calculated, expected), (
        f"Expected {expected} kg/s, but obtained {calculated} kg/s."
    )

In [8]:
def test_mass_flow_rate_zero_velocity():
    calculated = mass_flow_rate(
        rho=1000.0,
        area=0.02,
        normal_speed=0.0,
    )

    assert np.isclose(calculated, 0.0), (
        "Zero normal velocity must produce zero mass flow rate."
    )

In [9]:
def test_mass_flow_rate_rejects_invalid_inputs():
    invalid_cases = [
        (-1000.0, 0.02, 3.0, "negative density"),
        (0.0, 0.02, 3.0, "zero density"),
        (1000.0, -0.02, 3.0, "negative area"),
        (1000.0, 0.0, 3.0, "zero area"),
        (1000.0, 0.02, -3.0, "negative speed"),
    ]

    for rho, area, speed, case_name in invalid_cases:
        try:
            mass_flow_rate(rho, area, speed)

        except ValueError:
            continue

        raise AssertionError(
            f"The function did not reject {case_name}."
        )

In [10]:
def test_straight_duct_zero_net_momentum_flux():
    inlets = [
        {
            "velocity": [4.0, 0.0, 0.0],
            "mass_flow_rate": 10.0,
        }
    ]

    outlets = [
        {
            "velocity": [4.0, 0.0, 0.0],
            "mass_flow_rate": 10.0,
        }
    ]

    calculated = net_momentum_flux(inlets, outlets)
    expected = np.array([0.0, 0.0, 0.0])

    assert np.allclose(calculated, expected), (
        f"Expected {expected} N, but obtained {calculated} N."
    )

In [11]:
def test_accelerating_nozzle():
    inlets = [
        {
            "velocity": [2.0, 0.0, 0.0],
            "mass_flow_rate": 20.0,
        }
    ]

    outlets = [
        {
            "velocity": [5.0, 0.0, 0.0],
            "mass_flow_rate": 20.0,
        }
    ]

    calculated = net_momentum_flux(inlets, outlets)
    expected = np.array([60.0, 0.0, 0.0])

    assert np.allclose(calculated, expected), (
        f"Expected {expected} N, but obtained {calculated} N."
    )

In [12]:
def test_ninety_degree_elbow():
    inlets = [
        {
            "velocity": [4.0, 0.0, 0.0],
            "mass_flow_rate": 20.0,
        }
    ]

    outlets = [
        {
            "velocity": [0.0, 4.0, 0.0],
            "mass_flow_rate": 20.0,
        }
    ]

    calculated = net_momentum_flux(inlets, outlets)
    expected = np.array([-80.0, 80.0, 0.0])

    assert np.allclose(calculated, expected), (
        f"Expected {expected} N, but obtained {calculated} N."
    )

In [13]:
def test_symmetric_opposing_inlets():
     
    # This is an algebraic unit test for vector cancellation only. 
    # It does not represent a complete steady physical control-volume problem,
    # because no outlet or accumulation term is included.
    
    
    inlets = [
        {
            "velocity": [5.0, 0.0, 0.0],
            "mass_flow_rate": 3.0,
        },
        {
            "velocity": [-5.0, 0.0, 0.0],
            "mass_flow_rate": 3.0,
        },
    ]

    outlets = []

    calculated = net_momentum_flux(inlets, outlets)
    expected = np.array([0.0, 0.0, 0.0])

    assert np.allclose(calculated, expected), (
        f"Symmetric inlet momenta should cancel, but obtained {calculated} N."
    )

In [14]:
def test_multiple_inlets_and_outlets():
    inlets = [
        {
            "velocity": [3.0, 0.0, 0.0],
            "mass_flow_rate": 2.0,
        },
        {
            "velocity": [0.0, 4.0, 0.0],
            "mass_flow_rate": 1.0,
        },
    ]

    outlets = [
        {
            "velocity": [2.0, 1.0, 0.0],
            "mass_flow_rate": 3.0,
        }
    ]

    calculated = net_momentum_flux(inlets, outlets)

    # Outlet momentum:
    # 3(2, 1, 0) = (6, 3, 0)
    #
    # Inlet momentum:
    # 2(3, 0, 0) + 1(0, 4, 0) = (6, 4, 0)
    #
    # Net:
    # (6, 3, 0) - (6, 4, 0) = (0, -1, 0)

    expected = np.array([0.0, -1.0, 0.0])

    assert np.allclose(calculated, expected), (
        f"Expected {expected} N, but obtained {calculated} N."
    )

In [15]:
def test_uniform_velocity_shift_for_mass_balanced_flux():
    inlets_original = [
        {
            "velocity": [2.0, 0.0, 0.0],
            "mass_flow_rate": 10.0,
        }
    ]

    outlets_original = [
        {
            "velocity": [0.0, 2.0, 0.0],
            "mass_flow_rate": 10.0,
        }
    ]

    translation_velocity = np.array([7.0, -3.0, 2.0])

    inlets_shifted = [
        {
            "velocity": (
                np.asarray(stream["velocity"])
                + translation_velocity
            ).tolist(),
            "mass_flow_rate": stream["mass_flow_rate"],
        }
        for stream in inlets_original
    ]

    outlets_shifted = [
        {
            "velocity": (
                np.asarray(stream["velocity"])
                + translation_velocity
            ).tolist(),
            "mass_flow_rate": stream["mass_flow_rate"],
        }
        for stream in outlets_original
    ]

    original_flux = net_momentum_flux(
        inlets_original,
        outlets_original,
    )

    shifted_flux = net_momentum_flux(
        inlets_shifted,
        outlets_shifted,
    )

    assert np.allclose(original_flux, shifted_flux), (
        "For a mass-balanced steady flow, adding the same "
        "translation velocity to every stream should not alter "
        "the net momentum flux.\n"
        f"Original: {original_flux}\n"
        f"Shifted: {shifted_flux}"
    )

In [16]:
def test_relative_velocity_normal_case():
    flow_velocity = [10.0, 2.0, 0.0]
    control_surface_velocity = [4.0, -1.0, 0.0]

    calculated = relative_velocity(
        flow_velocity,
        control_surface_velocity,
    )

    expected = np.array([6.0, 3.0, 0.0])

    assert np.allclose(calculated, expected), (
        f"Expected {expected} m/s, but obtained {calculated} m/s."
    )

In [17]:
def test_relative_velocity_when_surface_moves_with_flow():
    flow_velocity = [5.0, 1.0, 0.0]
    control_surface_velocity = [5.0, 1.0, 0.0]

    calculated = relative_velocity(
        flow_velocity,
        control_surface_velocity,
    )

    expected = np.zeros(3)

    assert np.allclose(calculated, expected), (
        "A control surface moving with the fluid must have "
        "zero relative velocity."
    )

In [18]:
def test_relative_velocity_rejects_invalid_vectors():
    invalid_cases = [
        ([1.0, 2.0], [0.0, 0.0, 0.0]),
        ([1.0, 2.0, 3.0], [0.0, 0.0]),
        ([1.0, 2.0, 3.0, 4.0], [0.0, 0.0, 0.0]),
    ]

    for flow_velocity, surface_velocity in invalid_cases:
        try:
            relative_velocity(
                flow_velocity,
                surface_velocity,
            )

        except ValueError:
            continue

        raise AssertionError(
            "The function did not reject an invalid velocity vector."
        )

In [19]:
def test_net_momentum_flux_rejects_invalid_velocity_dimension():
    inlets = [
        {
            "velocity": [3.0, 0.0],
            "mass_flow_rate": 5.0,
        }
    ]

    outlets = []

    try:
        net_momentum_flux(inlets, outlets)

    except ValueError:
        return

    raise AssertionError(
        "A two-component velocity vector was not rejected."
    )

In [20]:
def test_net_momentum_flux_rejects_negative_mass_flow_rate():
    inlets = [
        {
            "velocity": [3.0, 0.0, 0.0],
            "mass_flow_rate": -5.0,
        }
    ]

    outlets = []

    try:
        net_momentum_flux(inlets, outlets)

    except ValueError:
        return

    raise AssertionError(
        "A negative mass flow rate was not rejected."
    )

In [21]:
def test_nonuniform_velocity_momentum_flux():
    rho = 1.0

    velocities = np.array([
        0.0,
        2.0,
        4.0,
        2.0,
        0.0,
    ])

    area_elements = np.full(
        shape=velocities.size,
        fill_value=0.2,
    )

    total_area = np.sum(area_elements)

    average_velocity = (
        np.sum(velocities * area_elements)
        / total_area
    )

    momentum_flux_from_profile = np.sum(
        rho * velocities**2 * area_elements
    )

    momentum_flux_from_average = (
        rho
        * total_area
        * average_velocity**2
    )

    beta = (
        momentum_flux_from_profile
        / momentum_flux_from_average
    )

    print(f"Average velocity: {average_velocity:.4f} m/s")
    print(
        "Momentum flux from velocity profile: "
        f"{momentum_flux_from_profile:.4f} N"
    )
    print(
        "Momentum flux from average velocity: "
        f"{momentum_flux_from_average:.4f} N"
    )
    print(f"Momentum correction factor beta: {beta:.4f}")

    assert momentum_flux_from_profile > momentum_flux_from_average, (
        "For this nonuniform profile, the profile-integrated "
        "momentum flux should exceed the estimate based only "
        "on average velocity."
    )

    assert beta > 1.0, (
        "The momentum correction factor should exceed 1 "
        "for this nonuniform velocity profile."
    )

In [22]:
def momentum_residual(
    external_forces,
    inlets,
    outlets,
):
    external_forces = np.asarray(
        external_forces,
        dtype=float,
    )

    if external_forces.shape != (3,):
        raise ValueError(
            "external_forces must contain exactly three components."
        )

    momentum_flux = net_momentum_flux(
        inlets,
        outlets,
    )

    return external_forces - momentum_flux

In [23]:
def test_nozzle_momentum_residual():
    inlets = [
        {
            "velocity": [2.0, 0.0, 0.0],
            "mass_flow_rate": 20.0,
        }
    ]

    outlets = [
        {
            "velocity": [5.0, 0.0, 0.0],
            "mass_flow_rate": 20.0,
        }
    ]

    external_forces = np.array([60.0, 0.0, 0.0])

    residual = momentum_residual(
        external_forces,
        inlets,
        outlets,
    )

    assert np.allclose(
        residual,
        np.zeros(3),
        atol=1e-12,
    ), (
        f"Momentum residual should be zero, but obtained {residual} N."
    )

In [24]:
def test_elbow_force_residual():
    inlets = [
        {
            "velocity": [4.0, 0.0, 0.0],
            "mass_flow_rate": 20.0,
        }
    ]

    outlets = [
        {
            "velocity": [0.0, 4.0, 0.0],
            "mass_flow_rate": 20.0,
        }
    ]

    pressure_force = np.array([600.0, -400.0, 0.0])
    wall_force = np.array([-680.0, 480.0, 0.0])

    total_external_force = pressure_force + wall_force

    residual = momentum_residual(
        total_external_force,
        inlets,
        outlets,
    )

    assert np.allclose(
        residual,
        np.zeros(3),
        atol=1e-12,
    ), (
        f"Elbow momentum residual should be zero, "
        f"but obtained {residual} N."
    )

In [25]:
tests = [
    test_mass_flow_rate_normal_case,
    test_mass_flow_rate_zero_velocity,
    test_mass_flow_rate_rejects_invalid_inputs,
    test_straight_duct_zero_net_momentum_flux,
    test_accelerating_nozzle,
    test_ninety_degree_elbow,
    test_symmetric_opposing_inlets,
    test_multiple_inlets_and_outlets,
    test_uniform_velocity_shift_for_mass_balanced_flux,
    test_relative_velocity_normal_case,
    test_relative_velocity_when_surface_moves_with_flow,
    test_relative_velocity_rejects_invalid_vectors,
    test_net_momentum_flux_rejects_invalid_velocity_dimension,
    test_net_momentum_flux_rejects_negative_mass_flow_rate,
    test_nonuniform_velocity_momentum_flux,
    test_nozzle_momentum_residual,
    test_elbow_force_residual,
]

results = [run_test(test) for test in tests]

passed = sum(results)
total = len(results)

print()
print("=" * 45)
print(f"Test summary: {passed}/{total} tests passed")
print("=" * 45)

if passed != total:
    raise AssertionError(
        f"{total - passed} test(s) failed."
    )

PASS: test_mass_flow_rate_normal_case
PASS: test_mass_flow_rate_zero_velocity
PASS: test_mass_flow_rate_rejects_invalid_inputs
PASS: test_straight_duct_zero_net_momentum_flux
PASS: test_accelerating_nozzle
PASS: test_ninety_degree_elbow
PASS: test_symmetric_opposing_inlets
PASS: test_multiple_inlets_and_outlets
PASS: test_uniform_velocity_shift_for_mass_balanced_flux
PASS: test_relative_velocity_normal_case
PASS: test_relative_velocity_when_surface_moves_with_flow
PASS: test_relative_velocity_rejects_invalid_vectors
PASS: test_net_momentum_flux_rejects_invalid_velocity_dimension
PASS: test_net_momentum_flux_rejects_negative_mass_flow_rate
Average velocity: 1.6000 m/s
Momentum flux from velocity profile: 4.8000 N
Momentum flux from average velocity: 2.5600 N
Momentum correction factor beta: 1.8750
PASS: test_nonuniform_velocity_momentum_flux
PASS: test_nozzle_momentum_residual
PASS: test_elbow_force_residual

Test summary: 17/17 tests passed
